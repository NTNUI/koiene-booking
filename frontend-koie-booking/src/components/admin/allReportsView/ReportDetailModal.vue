<template>
  <v-dialog :value="showModal" persistent width="500" @click:outside="closeModal">
    <v-card>
      <v-card-title class="headline">
        {{ title }}
      </v-card-title>
      <v-card-text>
        <slot name="modalContext" />
      </v-card-text>
      <v-card-actions>
        <v-spacer></v-spacer>
        <v-btn color="green darken-1" text @click="closeModal">
          Lukk
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script>
import Vue from 'vue';

export default Vue.extend({
  name: 'ReportDetailModal',
  props: {
    showModal: {
      type: Boolean,
      default: true,
    },
    closeModal: {
      type: Function,
      default() {
        return null;
      },
    },
    title: {
      type: String,
      default: '',
    },
  },
});
</script>

<style scoped></style>
