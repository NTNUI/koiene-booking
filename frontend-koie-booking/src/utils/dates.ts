import dayjs from 'dayjs';
require('dayjs/locale/nb');
dayjs.locale('nb');

export function getDateString(baseDate?: string, addDays?: number): string {
  let resultString: string;
  if (baseDate && addDays) {
    resultString = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else if (baseDate) {
    resultString = new Date(baseDate).toISOString().substr(0, 10);
  } else if (addDays) {
    resultString = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else {
    resultString = new Date().toISOString().substr(0, 10);
  }
  return resultString;
}

export function getDate(baseDate?: string, addDays?: number): Date {
  let resultDate: Date;
  if (baseDate && addDays) {
    resultDate = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000);
  } else if (baseDate) {
    resultDate = new Date(baseDate);
  } else if (addDays) {
    resultDate = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000);
  } else {
    resultDate = new Date();
  }
  return resultDate;
}

export function addToDate(startDate: string, howMany: number, timeUnit: dayjs.OpUnitType): string {
  const date = dayjs(startDate + 'T00:00:00.000Z');
  return date.add(howMany, timeUnit).format('YYYY-MM-DD');
}

export function formatDate(dateISO: string, formatString: string): string {
  const date = dayjs(dateISO + 'T00:00:00.000Z');
  return date.format(formatString);
}
