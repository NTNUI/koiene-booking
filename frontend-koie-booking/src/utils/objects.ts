export const clone = (object: any) => Object.assign({}, object);

export const renameKey = (object: any, key: string, newKey: string): any => {
  const value = object[key];
  delete object[key];
  object[newKey] = value;
  return object;
};

export const sortObjectByKey = (unordered: any) => {
  let ordered: any;
  ordered = {};
  Object.keys(unordered)
    .sort()
    .forEach(function(key) {
      ordered[key] = unordered[key];
    });
  return ordered;
};
