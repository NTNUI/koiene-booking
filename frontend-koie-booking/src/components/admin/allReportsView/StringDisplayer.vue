<template>
  <div :style="'cursor:' + (clickable ? 'pointer' : 'inherit')" @click="clickHandler">
    <div style="text-align: center; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; max-width: 150px">
      {{ text }}
    </div>
    <ReportDetailModal v-if="clickable" :title="title" :close-modal="closeModal" :show-modal="showModal">
      <template slot="modalContext">
        <p>{{ text }}</p>
      </template>
    </ReportDetailModal>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import ReportDetailModal from '@/components/admin/allReportsView/ReportDetailModal.vue';

export default Vue.extend({
  name: 'StringDisplayer',
  components: { ReportDetailModal },
  props: {
    text: {
      type: String,
      default: '',
    },
    clickable: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      showModal: false,
    };
  },
  methods: {
    closeModal() {
      this.showModal = false;
    },
    clickHandler() {
      if (this.clickable) this.showModal = true;
    },
  },
});
</script>
