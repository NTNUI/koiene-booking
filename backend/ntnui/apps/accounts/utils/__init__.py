from accounts.utils.mail_utils import send_gsuite_confirmation_mail
