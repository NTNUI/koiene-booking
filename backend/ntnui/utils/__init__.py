from ntnui.utils.pagination import PageSize7, PageSize20, PageSize100
