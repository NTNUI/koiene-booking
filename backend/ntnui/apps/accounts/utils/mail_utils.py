from django.core.mail import send_mail
from django.template.loader import render_to_string

from ntnui.enums import MembershipType


def new_membership(membership):
    """ Sends confirmation email to users and guests who signs up for events and sub-events. """

    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [membership.member.email]
    subject = "Welcome"

    context = {"user": membership.member, "group": membership.group}

    # Mail body.
    plain_message = render_to_string("emails/new_membership.txt", context)
    html_message = render_to_string("emails/new_membership.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_gsuite_confirmation_mail(user, contact_email, password):
    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = "Opprettelse av ny NTNUI-bruker"

    context = {
        "first_name": user.first_name,
        "email": user.email,
        "contact_email": contact_email,
        "password": password,
    }

    # Mail body.
    plain_message = render_to_string("emails/new_gsuite_account.txt", context)
    html_message = render_to_string("emails/new_gsuite_account.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_welcome_mail_to_leader(user, group):
    """ Send a mail to a new leader """
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny leder i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_leader_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_leader_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_cashier(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny kasserer i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_cashier_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_cashier_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_board_member(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny tillitsvalgt i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_board_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_board_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail(membership):
    """ Delegate welcome mail to different board members """
    send_welcome_mail_to_board_member(user=membership.member, group=membership.group)
    if membership.type == MembershipType.leader:
        send_welcome_mail_to_leader(user=membership.member, group=membership.group)

    if membership.type == MembershipType.cashier:
        send_welcome_mail_to_cashier(user=membership.member, group=membership.group)
