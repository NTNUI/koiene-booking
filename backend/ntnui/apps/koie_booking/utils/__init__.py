from koie_booking.utils.mail_utils import send_confirmation_mail, send_koie_information_mail
from koie_booking.utils.date_utils import get_daterange
