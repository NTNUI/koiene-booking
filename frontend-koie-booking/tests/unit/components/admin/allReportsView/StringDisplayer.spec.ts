import Vue from 'vue';
import Vuetify from 'vuetify';
import { createShallowWrapper } from '../../../utils';
import { Wrapper } from '@vue/test-utils';

import StringDisplayer from '@/components/admin/allReportsView/StringDisplayer.vue';
import ReportDetailModal from '@/components/admin/allReportsView/ReportDetailModal.vue';

Vue.use(Vuetify);

describe('Component StringDisplayer', () => {
  let wrapper: Wrapper<any>;

  beforeEach(() => {
    const wrapperOptions = {
      propsData: {
        text: 'Example text',
        clickable: true,
        title: 'Example title',
      },
      data() {
        return { showDialog: true };
      },
    };
    wrapper = createShallowWrapper(StringDisplayer, wrapperOptions);
  });

  it('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Closes dialog', () => {
    wrapper.vm.closeDialog();

    expect(wrapper.vm.showDialog).toBeFalsy();
  });

  it('Shows default values when no inputs are given', () => {
    wrapper = createShallowWrapper(StringDisplayer);
    expect(wrapper.vm.text).toEqual('');
    expect(wrapper.vm.clickable).toEqual(false);
    expect(wrapper.vm.title).toEqual('');
  });

  it('Click handler is called when StringDisplayer is clicked', () => {
    const clickHandler = jest.fn();
    const wrapperOptions = {
      propsData: {
        text: 'Example text',
        clickable: true,
        title: 'Example title',
      },
      methods: {
        clickHandler,
      },
    };
    wrapper = createShallowWrapper(StringDisplayer, wrapperOptions);
    const stringDisplayer = wrapper.find({ ref: 'stringDisplayerClickHandler' });
    stringDisplayer.trigger("click")

    expect(clickHandler).toHaveBeenCalled()
  });

  it('Does not show dialog when clickable is false', () => {
    const wrapperOptions = {
      propsData: {
        clickable: false,
      },
    };
    wrapper = createShallowWrapper(StringDisplayer, wrapperOptions);

    wrapper.vm.clickHandler();

    expect(wrapper.vm.showDialog).toBeFalsy();
  });

  it('Shows dialog when clickable is true', () => {
    wrapper.vm.clickHandler();

    expect(wrapper.vm.showDialog).toBeTruthy();
  });
});
