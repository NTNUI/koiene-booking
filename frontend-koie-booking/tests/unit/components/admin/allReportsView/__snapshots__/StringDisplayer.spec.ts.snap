// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component StringDisplayer Matches snapshot 1`] = `
<div style="cursor: pointer;">
  <div style="text-align: center; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; max-width: 150px;">
    Example text
  </div>
  <reportdetailmodal-stub showmodal="true" closemodal="function () { [native code] }" title="Example title"><template></template></reportdetailmodal-stub>
</div>
`;
